configuration - add set operations union, intersection, difference

albinsjolund opened the pull request:
Added options to performe set operations on configuration For example:
casual domain --configuration-get | casual configuration --union new_config.yaml

lazan commented on middleware/administration/unittest/source/cli/configuration/test_set_operations.cpp:
```diff
@@ -0,0 +1,145 @@
+      - alias: Server3
+)");
+
+      auto command = "casual domain --configuration-get | casual configuration --union " + rhs.string();
```
What if:  `casual domain --configuration-get json | ... `?

albinsjolund replied:
I have only tested that manually and it works, but I will add a testcase for that too.

lazan replied:
How can it work?

albinsjolund replied:
I don't know. Json works but not xml

lazan commented on middleware/administration/unittest/source/cli/configuration/test_set_operations.cpp:
```diff
@@ -0,0 +1,145 @@
+      // Expect to find all 3 servers
+      EXPECT_TRUE( union_output.find("Server1") != std::string::npos) <<  union_output;
+      EXPECT_TRUE( union_output.find("Server2") != std::string::npos) <<  union_output;
+      EXPECT_TRUE( union_output.find("Server3") != std::string::npos) <<  union_output;
```
We should use something else than std::string index based "algorithms". Index based are more error prune than iterator based (at least as I see it), especially if "mixed" throughout the code base. 
I don't think this specific case is that bad though, it's more a principle...

albinsjolund replied:
Do you mean like this? `union_output.find( "Server1") < union_output.length()`
Or is it `find` I should not use? What other options do I have?

lazan replied:
More like `EXPECT_TRUE( algoritm::find( union_output, "Server1"))`... Not sure if it will work with string literal, you might have to use string_view. `"Server1"sv`

Or, keep it as if.

lazan commented on middleware/configuration/unittest/source/test_model.cpp:
```diff
@@ -454,6 +454,237 @@ namespace casual
+         auto B = local::configuration( remove);
+         auto result = local::configuration( expected);
+
+         set_difference( A, std::move( B));
```
I'm not sure if I like that set_xxx are mutating... Feels wrong to update A in this case... To me, it would be more _functional_ to have `A = set_difference( std::move( A), std::move( B));` Or, if one need to keep originals: `auto result = set:difference( A, B);`

lazan commented on middleware/configuration/unittest/source/test_model.cpp:
```diff
@@ -454,6 +454,237 @@ namespace casual
+         set_intersection( A, std::move( B));
+
+         EXPECT_TRUE( A == result) << CASUAL_NAMED_VALUE( A) << "\n" << CASUAL_NAMED_VALUE( result);
+      }
```
Nice set of test cases. Feels like we've got good coverage.

albinsjolund committed:
configuration - add set operations union, intersection, difference

Added options to performe set operations on configuration
For example:
casual domain --configuration-get | casual configuration --union new_config.yaml
